addon/utils/disposable.js

```diff
@@ -0,0 +1,89 @@
+import Ember from 'ember';
+import { assert } from '@ember/debug';
+import { WILL_DESTROY_PATCHED } from '../utils/flags';
+
+const { WeakMap } = Ember;
+
+/**
+ * A map of instances/array of disposables. Only exported for
+ * testing purposes.
+ *
+ * @public
+ */
+export let registeredDisposables = new WeakMap();
+
+/**
+ * Registers a new disposable function to run for an instance. Will
+ * handle lazily creating a new array to store the disposables per
+ * instance if one does not exist. Will additionally patch an object's
+ * `willDestroy` hook to ensure the destroyables are run/destroyed when
+ * the object is destroyed.
+ *
+ * @public
+ * @param {*} obj the instance to store the disposable for
+ * @param {*} dispose a function that disposes of instance resources
+ * @returns a disposable object, which wraps the passed in `dispose` function
+ */
+export function registerDisposable(obj, dispose) {
+  assert(
+    'Called `registerDisposable` where `obj` is not an object',
+    typeof obj === 'object'
+  );
+  assert(
+    'Called `registerDisposable` where `dispose` is not a function',
+    typeof dispose === 'function'
+  );
+  assert(
```

Comment:
this seems to just check that we are running `registerDisposable` on the mixin, if that is the intent we should test that directly. If not, then this seems somewhat confusing unless the functional style has a "setup disposable" in the constructor

Reply:
See my [comment](https://github.com/rwjblue/ember-lifeline/pull/63#discussion_r166052030) above. This isn't checking we're running `registerDisposable`, but checking that _any_ class or component is calling `runDisposables`, which ensures that tasks are correctly canceled. 

I believe we still need and want this.